### PR TITLE
Fix ambiguous Dumpable behaviour

### LINKED ISSUE #13733: Clicking a table with a pill canister is ambiguous
## Description

When clicking on a table using a partially filled pill canister while the table already has some pills on it, your character attempts to both dump the canister and pick up the pills already on the table at the same time. 

Additionally, the amount of pills that are dumped/picked up seems to vary depending on which action finishes first. If the canister picks up before it dumps, then the entire container is dumped, but if it's the other way around, then it dumps what was in the container and then picks up what it could *before it dumped*, leaving the pills that were just dumped untouched. This behaviour doesn't really make sense to me.

**Reproduction**

1. Obtain a pill canister filled with any type of pill.
2. Place one or more pills from the bottle on an empty table.
3. Left-click on the surface of the table while holding the partially empty pill canister.

**Screenshots**

dumping

**Additional context**

This issue probably exists for Dumpable Storage containers in general, such as plant bags etc., but I haven't tested these.

## Possible solutions

I can think of a few alternative solutions for when there's a conflict between pickup and dump, but would like some input on it before I submit a PR, since it would be my first PR on this project. 

1. Prioritize pickup over dump
2. Prioritize dump over pickup
3. Add a boolean `prioritizeOverAreaInsert` flag to the Dumpable component
3. Don't do anything
4. Give a dialog option every time there's a conflict

Personally, I prefer option 1, since you can always click again to dump in case you did an area pickup on accident, but you can't easily pick up the things you just dumped if you dumped on accident. Even though the most common use case for the pill canister is to dump its contents onto the counter, that might not be the case for other dumpable storages, as I could imagine you perform area inserts more often with a partially full plant bag than you dump, so for the component to be as generic as possible, we should go with the least annoying option.

And regarding option 3, I'm not sure whether it's actually better to make this prioritization vary depending on whether you're using a plant bag or a pill canister, since it could just as well be even more confusing for the user to have to learn which container does what. Therefore, the best thing might just be to just pick one option and stick with it.

Additionally, it would be nice to be able perform the other operation, eg. dump for option 1, with some kind of modifier, such as by holding alt. I'm not sure whether this is possible with the current interaction system though (there is no `AltInteract` field on the `AfterInteract` event), but if you know how, please let me know!

### PR DESCRIPTION
**Pull request by joelhed**
Fixes #13733 by prioritizing area insert over dumping when they conflict on `AfterInteract`.

## About the PR

**Media**

dumping-fixed

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**

:cl:
- fix: Fixed weird behaviour when clicking on a table using a dumpable storage item.

### CODE REVIEW

**Review by AJCM-git — CHANGES_REQUESTED** [review]
This makes the bug where you cant dump things on a table with non-whitelisted items above more obvious since the dumping on the table doesnt happen at the same time anymore

**Comment by joelhed** [comment]
Oh, I didn't know about that issue! Could you elaborate on how to reproduce it?

**Comment by AJCM-git** [comment]
Place like two things in a trashbag, place a non-trash entity above a table, then try clicking the table with the trashbag and the dump will not trigger as you "cant insert" said entity. 
This also happens in master but your changes make it more obvious as it fixes the dump that happens at the same time

**Comment by joelhed** [comment]
I fixed that issue too, so now you should be able to pick up trash on tables without any issue!

pickup-error-message-fixed

File: Content.Server/Storage/EntitySystems/StorageSystem.cs [hunk 1]
```diff
@@ -316,7 +316,8 @@ private async void AfterInteract(EntityUid uid, ServerStorageComponent storageCo
                 {
                     if (entity == args.User
                         || !itemQuery.HasComponent(entity)
-                        || !_interactionSystem.InRangeUnobstructed(args.User, entity))
+                        || !_interactionSystem.InRangeUnobstructed(args.User, entity)
+                        || !CanInsert(uid, entity, out _, storageComp))
```
Comment (metalgearsloth) [inline Content.Server/Storage/EntitySystems/StorageSystem.cs:320]: You should do the expensive checks last.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Content.Server/Storage/EntitySystems/StorageSystem.cs (hunk 1 above)
Comment (joelhed) [reply Content.Server/Storage/EntitySystems/StorageSystem.cs:320]: Ah, didn't think that InRangeUnobstructed would be more expensive, but yeah, all that raycasting couldn't possibly be faster than checking the existence of a couple of components. Thank you!

**Comment by joelhed** [comment]
I am considering changing the "Can't insert" message that appears on `InsertItemInWorld` more verbose as well by passing along the reason from `CanInsert` as an out parameter of `Insert`, so the new signature would become 

```C#
public bool Insert(EntityUid uid, EntityUid insertEnt, out string? reason, ServerStorageComponent? storageComp = null, bool playSound = true)
```

This would touch some more files that are using `Insert`, seeing as you can't have optional out parameters, and could be a little out of scope for this PR, but if you think it's a good idea I'll do it. 

I guess you could just call `CanInsert` again in `InsertItemInWorld`, but it seems a little unnecessary to add even more CanInsert calls than is necessary. But you guys are the experts here, what would you do?

**Comment by AJCM-git** [comment]
the message is good enough for now and gets the point across, it can be improved in the future if you really want to

**Review by AJCM-git — APPROVED** [review]

**Pull request merged**